### PR TITLE
Go: add showcase tests

### PR DESCRIPTION
**Pull request by landrito**

### CODE REVIEW

File: showcase/go/showcase_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	clientOpt := option.WithGRPCConn(conn)
+	client, err = showcase.NewEchoClient(context.Background(), clientOpt)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	os.Exit(m.Run())
+}
+
+func TestEcho(t *testing.T) {
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: Consider `t.Parallel()`

File: showcase/go/showcase_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+		},
+	}
+	resp, err := client.Echo(context.Background(), req)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.GetContent() != req.GetContent() {
+		t.Errorf(
+			"Echo did not receive expected value. Got %s, Wanted %s",
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: Not a serious issue, though the messages can be shorter like `Echo() = %q, want %q`.

I'd rather use `%q`. It's unlikely but error messages with `%s` will be really confusing if we have unprintable characters.

File: showcase/go/showcase_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	val := codes.Canceled
+	req := &genprotopb.EchoRequest{
+		Response: &genprotopb.EchoRequest_Error{
+			Error: &spb.Status{Code: int32(val)},
+		},
+	}
+	_, err := client.Echo(context.Background(), req)
+
+	if err == nil {
+		t.Errorf("Echo called with code %d did not return an error.", val)
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: `t.Errorf("Echo() = %v, want error", resp)`? The response value could help figuring out what went wrong?

File: showcase/go/showcase_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	}
+	status, _ := status.FromError(err)
+	if status.Code() != val {
+		t.Errorf("Echo called with code %d returned an error with code %d", val, status.Code())
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: `Echo errors with %d, want %d`. Test errors idiomatically says "want" at the end.

File: showcase/go/showcase_integration_test.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+			break
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		resps = append(resps, resp.GetContent())
+	}
+	got := strings.Join(resps, " ")
+	if content != got {
+		t.Errorf("Expand expected %s but got %s", content, got)
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: "want" at the end.

File: showcase/go/showcase_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+		t.Fatal(err)
+	}
+
+	for _, str := range strings.Split(content, " ") {
+		s.Send(&genprotopb.EchoRequest{
+			Response: &genprotopb.EchoRequest_Content{Content: str}})
+	}
+
+	resp, err := s.CloseAndRecv()
+	if content != resp.GetContent() {
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: Check error.

File: showcase/go/showcase_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		}
+		resps = append(resps, resp.GetContent())
+	}
+	got := strings.Join(resps, " ")
+	if content != got {
+		t.Errorf("Chat expected %s but got %s", content, got)
+	}
+}
+
+func TestWait(t *testing.T) {
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: Do we need to test the case with timeout as well?

File: showcase/go/showcase_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	iter := client.Pagination(context.Background(), req)
+
+	expected := int32(0)
+	for {
+		i, err := iter.Next()
+		if err != nil {
+			break
+		}
+		if i != expected {
+			t.Errorf("Pagination expected val %d, got %d", expected, i)
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: want at the end.

File: showcase/go/showcase_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func TestPagination(t *testing.T) {
+	req := &genprotopb.PaginationRequest{PageSize: 5, MaxResponse: 20}
+	iter := client.Pagination(context.Background(), req)
+
+	expected := int32(0)
+	for {
+		i, err := iter.Next()
+		if err != nil {
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: I think we should check for `iterator.Done`

**Comment by andreamlin** [comment]
Does this need a "Test Showcase" job, like [Node does](https://circleci.com/gh/googleapis/gapic-generator/5480)?

**Comment by landrito** [comment]
Ah yes I was gonna ask about that.

The mock tests that were generated are failing on the client streaming methods. I suspect those tests were never tested since there are no real APIs that are using them.

This method is getting the following error `panic: runtime error: index out of range [recovered] panic: runtime error: index out of range` at the line I specified in comments.
```
func TestEchoCollect(t *testing.T) {
	var content string = "content951530617"
	var expectedResponse = &genprotopb.EchoResponse{
		Content: content,
	}

	mockEcho.err = nil
	mockEcho.reqs = nil

	mockEcho.resps = append(mockEcho.resps[:0], expectedResponse)

	var request *genprotopb.EchoRequest = &genprotopb.EchoRequest{}

	c, err := NewEchoClient(context.Background(), clientOpt)
	if err != nil {
		t.Fatal(err)
	}

	stream, err := c.Collect(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	resp, err := stream.CloseAndRecv()

	if err != nil {
		t.Fatal(err)
	}

        // The panic is being caused by this line.
	if want, got := request, mockEcho.reqs[0]; !proto.Equal(want, got) {
		t.Errorf("wrong request %q, want %q", got, want)
	}

	if want, got := expectedResponse, resp; !proto.Equal(want, got) {
		t.Errorf("wrong response %q, want %q)", got, want)
	}
}
```

This error method is simply not using the request.
```
func TestEchoCollectError(t *testing.T) {
	errCode := codes.PermissionDenied
	mockEcho.err = gstatus.Error(errCode, "test error")

	var request *genprotopb.EchoRequest = &genprotopb.EchoRequest{}

	c, err := NewEchoClient(context.Background(), clientOpt)
	if err != nil {
		t.Fatal(err)
	}

	stream, err := c.Collect(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	resp, err := stream.CloseAndRecv()

	if st, ok := gstatus.FromError(err); !ok {
		t.Errorf("got error %v, expected grpc error", err)
	} else if c := st.Code(); c != errCode {
		t.Errorf("got error code %q, want %q", c, errCode)
	}
	_ = resp
}
```

File: .circleci/config.yml [hunk 10]
```diff
@@ -584,10 +585,7 @@ jobs:
           command: |
             cd /tmp/workspace/gapic-generator/artman-genfiles
             for d in gapi-cloud-*; do
-              cp -r $d/cloud.google.com/ $GOPATH/src
-            done
-            for d in go/*; do
-              cp -r $d/google.golang.org $GOPATH/src
+              cp -r $d/* $GOPATH/src
```
[the remark this answers is not in the document]
Comment (landrito) [reply .circleci/config.yml]: Done!

File: showcase/go/showcase_integration_test.go (hunk 1 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: IIUC adding t.Parallel() to every test. Done!

File: showcase/go/showcase_integration_test.go (hunk 2 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Done!

File: showcase/go/showcase_integration_test.go (hunk 3 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Done!

File: showcase/go/showcase_integration_test.go (hunk 4 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Thanks! I've been trying to figure out what the idiomatic way to write these test errors for a while. Done!

File: showcase/go/showcase_integration_test.go (hunk 6 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Done!

File: showcase/go/showcase_integration_test.go (hunk 7 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Done!

File: showcase/go/showcase_integration_test.go (hunk 8 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Done!

File: showcase/go/showcase_integration_test.go (hunk 9 above)
Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Done!

File: showcase/go/showcase_integration_test.go [hunk 11]
```diff
@@ -109,11 +111,12 @@ func TestExpand(t *testing.T) {
 	}
 	got := strings.Join(resps, " ")
 	if content != got {
-		t.Errorf("Expand expected %s but got %s", content, got)
+		t.Errorf("Expand() = %q, want %q", got, content)
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: `content, got`?

File: showcase/go/showcase_integration_test.go [hunk 12]
```diff
@@ -172,25 +180,44 @@ func TestWait(t *testing.T) {
...
+	req := &genprotopb.WaitRequest{
+		ResponseDelay: &durationpb.Duration{Seconds: 1},
+		Response: &genprotopb.WaitRequest_Success{
+			Success: &genprotopb.WaitResponse{Content: content},
+		},
+	}
+	ctx, _ := context.WithTimeout(context.Background(), 20*time.Millisecond)
+	resp, _ := client.Wait(ctx, req)
+
+	if resp != nil {
```
Comment (pongad) [inline showcase/go/showcase_integration_test.go]: I'd rather check for error, not resp

Comment (landrito) [reply showcase/go/showcase_integration_test.go]: Done!

**Comment by landrito** [comment]
Thanks for the review! Filed #2347 in response to the broken client streaming mock test.

**Pull request merged**